northd/multicast.dl

```diff
@@ -0,0 +1,161 @@
+                                           true),
+        .flood_unreg    = map_get_bool_def(other_config,
+                                           "mcast_flood_unregistered",
+                                           false),
+        .eth_src        = map_get_str_def(other_config, "mcast_eth_src", ""),
+        .ip4_src        = map_get_str_def(other_config, "mcast_ip4_src", ""),
+        .table_size     = map_get_int_def(other_config,
+                                          "mcast_table_size",
+                                          mCAST_DEFAULT_MAX_ENTRIES()),
+        .idle_timeout   = idle_timeout,
+        .query_interval = query_interval,
+        .query_max_resp = query_max_resp) :-
+    nb.Logical_Switch(._uuid        = ls_uuid,
+                      .other_config = other_config),
+    var idle_timeout =
+        map_get_int_def_limit(other_config, "mcast_idle_timeout",
+                              mCAST_DEFAULT_IDLE_TIMEOUT_S(),
+                              mCAST_DEFAULT_MIN_IDLE_TIMEOUT_S(),
+                              mCAST_DEFAULT_MAX_IDLE_TIMEOUT_S()),
+    var query_interval =
+        map_get_int_def_limit(other_config, "mcast_query_interval",
+                              idle_timeout / 2,
+                              mCAST_DEFAULT_MIN_QUERY_INTERVAL_S(),
+                              mCAST_DEFAULT_MAX_QUERY_INTERVAL_S()),
+    var query_max_resp =
+        map_get_int_def(other_config, "mcast_query_max_response",
+                        mCAST_DEFAULT_QUERY_MAX_RESPONSE_S()).
+
+/* IP Multicast per router configuration. */
+relation &McastRouterCfg(
+    datapath: uuid,
+    relay   : bool
+)
+
+&McastRouterCfg(lr_uuid, mcast_relay) :-
+    nb.Logical_Router(._uuid = lr_uuid, .options = options),
+    var mcast_relay = map_get_bool_def(options, "mcast_relay", false).
+
+/* Mapping between Switch and the set of router port uuids on which to flood
+ * IP multicast for relay.
+ */
+relation SwitchMcastFloodRelayPorts(sw: Ref<Switch>, ports: Set<uuid>)
+
+SwitchMcastFloodRelayPorts(switch, relay_ports) :-
+    &SwitchPort(
+        .lsp  = lsp,
+        .sw = switch,
+        .peer = Some{&RouterPort{.router = &Router{.mcast_cfg = &mcast_cfg}}}
+    ), mcast_cfg.relay,
+    var lsp_uuid = lsp._uuid,
+    var relay_ports = Aggregate((switch), group2set(lsp_uuid)).
+
+SwitchMcastFloodRelayPorts(switch, set_empty()) :-
+    Switch[switch].
+
+/* Relation used to determine if a Switch has at least one router port
+ * connected to a Router with multicast relay enabled.
+ */
+relation SwitchHasMcastFloodRelay(sw: Ref<Switch>, has_flood_relay: bool)
+
+SwitchHasMcastFloodRelay(switch, not set_is_empty(ports)) :-
+    SwitchMcastFloodRelayPorts(switch, ports).
+
+/* Flattened IGMP group. One record per address-port tuple. */
+relation IgmpSwitchGroupPort(
+    address: string,
+    switch : Ref<Switch>,
+    port   : uuid
+)
+
+IgmpSwitchGroupPort(address, switch, lsp._uuid) :-
```

Comment:
> This equivalent rule should be more efficient, as it will only consider `SwitchPort` records that match `pb_port_uuid`:
> 
> ```
> IgmpSwitchGroupPort(address, switch, lsp_uuid) :-
>     sb.IGMP_Group(.address = address, .datapath = igmp_dp_set,
>                   .ports = pb_ports),
>     var pb_port_uuid = FlatMap(pb_ports),
>     sb.Port_Binding(._uuid = pb_port_uuid, .logical_port = lsp_name),
>     &SwitchPort(.lsp = nb.Logical_Switch_Port{._uuid = lsp_uuid, .name = lsp_name}, .sw = switch).
> ```

Thanks! done.